Docker-compose dev environment

hornc opened the pull request:
Closes #1063 
makes some changes related to ;) #1064 

and hopefully #1065

## Description:
* Clarifies the split between `olbase` and `oldev` 
    * `olbase` is a single application standard image that will run OL as the `openlibrary` user, needs to be deployed sensibly to be useful
    * `oldev` has whatever else we need to operate our dev environment, i.e. pre-populated dev db, other services and tools. `oldev` may, and _should_, be reduced over time as we split out services.

* Volume mounts local dev directory and creates docker compose volumes to retain the js + css that was created at build time, and the `vendor` dir that is used to fetch git submodules, which is also populated on build.

* Uses a new `openlibray-docker.yml` config file to not interfere with vagrant, and avoids using `sed` to fix things. I'd suggested we rename this back to `openlibrary.yml` when we fully remove all traces of vagrant.

* Adds the gunicorn `--reload` option in dev so code changes are automatically detected and reloaded :tada: 

## Usage
With the volume mount in place and the gunicorn  `--reload` option a developer can edit code locally and the webserver will automatically reload the code on save.

The instructions in docker/README.md are :ok_hand:

mekarpeles commented on conf/openlibrary-docker.yml:
```diff
@@ -0,0 +1,208 @@
+    editions_view: http://127.0.0.1:5984/editions/_fti/_design/seeds/by_seed
+
+admin:
+    olsystem_root: /home/noufal/projects/OL/olsystem/
```
I think this can be pruned?

hornc replied:
It is used in code here: https://github.com/internetarchive/openlibrary/blob/75452cf6f0e3c23e1030395c93e232686ebaa417/openlibrary/plugins/admin/code.py#L538  but I can't get the page `/admin/services` to display. Looks like it is not registered in the `setup()` further down in that file.   Shall I prune the entire `service_status` class?

mekarpeles commented on conf/openlibrary-docker.yml:
```diff
@@ -0,0 +1,208 @@
+
+libraries_admin_email: admin@dev-instance
+
+graphite_base_url: http://ol-admin.us.archive.org:7080
```
I don't think we want developer instances recording to graphite (analytics)

hornc commented on conf/openlibrary-docker.yml:
```diff
@@ -0,0 +1,208 @@
+libraries_admin_email: admin@dev-instance
+
+graphite_base_url: http://ol-admin.us.archive.org:7080
+geoip_database: var/cache/GeoLiteCity.dat
```
@mekarpeles what is this for, I don't see it in olsystem?

hornc replied:
Used in https://github.com/internetarchive/openlibrary/blob/75452cf6f0e3c23e1030395c93e232686ebaa417/openlibrary/core/geo_ip.py

mekarpeles commented:
lgmt, requesting one more pair of eyeballs; there's a lot here. Config especially looks much better

cdrini commented:
👍 I'll take a look/test it out over the weekend! These changes look great, btw :)

cdrini commented on conf/openlibrary-docker.yml:
```diff
@@ -0,0 +1,173 @@
+# Open Library Docker (dev) configuration
```
The removal of unnecessary/outdated keys and whatnot should also be reflected in the original config file, for consistency (although it will be deleted once we switch to docker :P)

cdrini commented on docker/README.md:
```diff
@@ -33,8 +33,7 @@ docker-compose stop
 docker-compose rm -v
```
Can you add a section about "Code updates", describing the work flow when making changes a) python code (do nothing); b) templates (do nothing); c) frontend code (`docker-compose exec web make css js`).

cdrini commented:
I'm not sure if it's because of this PR, but #1051 also no longer appears to be an issue \o/

hornc commented on docker/ol-docker-start.sh:
```diff
@@ -3,6 +3,17 @@
 
+CONFIG=conf/openlibrary-docker.yml
+
+reindex-solr() {
```
Agree, I actually tried to look into doing this, but didn't think I could do it quickly without risking breaking all the other make tasks. I decided we already had a big limitation that the Makefile assumed hardcoded solr locations for tasks and that correcting it would be a full feature. I also think that most of that Makefile needs a review to check we even want to have all of those tasks there.   I have strong doubts about more than half of them, they are either vagrant only, not used, or possibly broken, but I'm not sure :)

I feel there should be an independent "review Makefile" issue, and more discussion. I also wasn't sure if putting functions with loops in Makefiles was nice. Perhaps the reindex task should be a script that is called from the Makefile?  (I know you're not a fan of :all-the-things: in '/scripts' but I'm not sure the Makefile is right for it either ;) )

cdrini replied:
The makefile definitely needs some love :P if it's more work than a) add env variable to docker-compose, b) add env variable to vagrant, c) replace hardcoded config file in Makefile with said variable, then I wouldn't worry about it. Certainly not mission critical, and it'll be easy to clean up once we replace vagrant.

(Also, the current implementation in the Makefile doesn't use loops, just has to (repetitive) commands. Actually, since this uses a regex to query psql, it could be done on just one command...)

hornc replied:
oh wow, that's heaps better in one command :man_facepalming:  I thought my loop was better than the duplication, not really. I'll revisit!

cdrini replied:
Oh, drats, I'm wrong; I think maybe their order matters (because one gets queried for the other) :/

hornc replied:
I think I'll leave it as is for now, if that's ok?

hornc commented:
@cdrini if this fixes #1051, that is great!  Keep an eye on it, because even though I think I see how this might have fixed it (some of those dirs will no longer be accessing the windows fs, but using the docker volume mounts) I think some of the symlinks will still be mounted from Windows. Test that running`make all` without your workaround doesn't error too. I think there a still some symlinks that will be taken from your local windows dir. At least if tghey are not all fixed, this suggests an approach to move them off the  Windows fs.

hornc commented:
@cdrini I think I have addressed all of your comments! Last one was the comments explaining the volume and bind mounts :+1:  :bowing_man:  Let me know if its good to go!

mekarpeles commented:
w/ latest changes, this lgtm

cdrini commented on conf/openlibrary.yml:
```diff
@@ -1,55 +1,32 @@
-# Open Library fastcgi configuration
+# Open Library Docker (dev) configuration
```
Assuming this should say vagrant? :P
